README.md
# Webcryptobox
WebCrypto compatible encryption with Bash and OpenSSL.

This package implements the [Webcryptobox](https://github.com/jo/webcryptobox) encryption API.

Compatible packages:
* [Webcryptobox Rust](https://github.com/jo/webcryptobox-rs)
* [Webcryptobox Bash](https://github.com/jo/webcryptobox-sh)

There is also a CLI tool: [wcb.sh](https://github.com/jo/wcb-js)

Demo: https://jo.github.io/webcryptobox-js/

This library provides easy to use and convenient wrappers around the WebCrypto primitives, as well as some helpers for encoding/decoding and a few sugar functions, with zero dependencies.

It works directly in the browser and in latest Node.js versions (via [the experimental WebCrypto API](https://nodejs.org/api/webcrypto.html)).


## Usage
In Node, you can use the lib as usual:
```sh
npm install webcryptobox
```

and then
```js
import * as wcb from 'webcryptobox'
```

In modern browser which support es6 modules, just include [the file](./webcryptobox.js) directly:
```html
<script type=module>
  import * as wcb from './webcryptobox.js'
</script>
```

Now you can dance with Webcryptobox like this:
```js
const alice = await wcb.generateKeyPair()
const bob = await wcb.generateKeyPair()
const text = 'Nobody else can offer me something, something heart felt like you did it.'
const message = wcb.decodeText(text)
const box = await wcb.encryptTo({ message, alice.privateKey, bob.publicKey })
const decryptedBox = await wcb.decryptFrom({ box, bob.privateKey, alice.publicKey })
const decryptedText = wcb.encodeText(decryptedBox)
```


## API
This lib is written with some ECMAScript 6 features, mainly modules, dynamic import, async, destructuring and object spreading.

Most of the functions return promises, except for the encoding/decoding utilities.


#### `cipher`
Returns a cipher identifier:

```js
wcb.cipher
// 'ECDH-P-521-AES-256-CBC'
```


### Utils for Encoding & Decoding
Webcryptobox provides utility functions to convert between several text representations and the internally used `Uint8Array`s.

#### `decodeText`
Takes an unicode string and encodes it to an Uint8Array:

```js
const data = wcb.decodeText('my message')
// Uint8Array(10) [
//   109, 121,  32, 109,
//   101, 115, 115,  97,
//   103, 101
// ]
```

#### `encodeText`
Given a Uint8Array, encodes the data as unicode string:

```js
const text = wcb.encodeText(new Uint8Array([
  109, 121,  32, 109,
  101, 115, 115,  97,
  103, 101
]))
// my message
```

#### `decodeHex`
Takes a hex string and encodes it to an Uint8Array:

```js
const data = wcb.decodeHex('6d79206d657373616765')
// Uint8Array(10) [
//   109, 121,  32, 109,
//   101, 115, 115,  97,
//   103, 101
// ]
```

#### `encodeHex`
Given a Uint8Array, encodes the data as hex string:

```js
const hex = wcb.encodeHex(new Uint8Array([
  109, 121,  32, 109,
  101, 115, 115,  97,
  103, 101
]))
// 6d79206d657373616765
```

#### `decodeBase64`
Takes a base64 string and encodes it to an Uint8Array:

```js
const data = wcb.decodeBase64('bXkgbWVzc2FnZQ==')
// Uint8Array(10) [
//   109, 121,  32, 109,
//   101, 115, 115,  97,
//   103, 101
// ]
```

#### `encodeBase64`
Given a Uint8Array, encodes the data as base64 string:

```js
const base64 = wcb.encodeBase64(new Uint8Array([
  109, 121,  32, 109,
  101, 115, 115,  97,
  103, 101
]))
// bXkgbWVzc2FnZQ==
```


### Key Generation and Derivation
Functions for generating a ecdh and aes-cbc keys, for deriving an aes-cbc key or the public key from a private one and for generating a sha-256 fingerprint of a key.

#### `generateKey`
Generate aes-cbc key with a length of 256. The key will be extractable and can be used for encryption and decryption.

```js
const key = await wcb.generateKey()
// CryptoKey {
//   type: 'secret',
//   extractable: true,
//   algorithm: { name: 'AES-GCM', length: 256 },
//   usages: [ 'encrypt', 'decrypt' ]
// }
```
#### `generateKeyPair`
Generates ecdh key pair with curve `P-521`. The private key will be extractable, and can be used to derive a key.

```js
const keyPair = await wcb.generateKeyPair()
// {
//   publicKey: CryptoKey {
//     type: 'public',
//     extractable: true,
//     algorithm: { name: 'ECDH', namedCurve: 'P-521' },
//     usages: []
//   },
//   privateKey: CryptoKey {
//     type: 'private',
//     extractable: true,
//     algorithm: { name: 'ECDH', namedCurve: 'P-521' },
//     usages: [ 'deriveKey' ]
//   }
// }
```

#### `getPublicKey`
Given a private key, returns its corresponding public key. As there is no direct API for this in WebCrypto, this utilizes import and export of the key (in `jwk` format), while removing the private key parts.

```js
const { privateKey } = await wcb.generateKeyPair()
const publicKey = await wcb.derivePublicKey(privateKey)
// CryptoKey {
//   type: 'public',
//   extractable: true,
//   algorithm: { name: 'ECDH', namedCurve: 'P-521' },
//   usages: []
// }
```

#### `deriveKey`
Given a private and a public key, this function derives an aes-cbc key. For two key pairs `A` and `B`, the derived key will be the same for
`deriveKey({ privateKey: A.privateKey, publicKey: B.publicKey })` and `deriveKey({ privateKey: B.privateKey, publicKey: A.publicKey })`.

```js
const { privateKey } = await wcb.generateKeyPair()
const { publicKey } = await wcb.generateKeyPair()
const key = await wcb.deriveKey({ privateKey, publicKey })
// CryptoKey {
//   type: 'secret',
//   extractable: true,
//   algorithm: { name: 'AES-GCM', length: 256 },
//   usages: [ 'encrypt', 'decrypt' ]
// }
```

#### `deriveBits`
Given a key pair, this function derives 16 bytes in an ArrayBuffer.

```js
const { privateKey, publicKey } = await wcb.generateKeyPair()
const key = await wcb.deriveBits({ privateKey, publicKey })
// ArrayBuffer {
//   [Uint8Contents]: <ac 54 d5 01 74 ca d6 87 f5 65 18 d0 4f e4 0f 18 77 ... more bytes>,
//   byteLength: 16
// }
```

#### `derivePassword`
Given a key pair, this function derives a password of given length as an ArrayBuffer.

```js
const { privateKey, publicKey } = await wcb.generateKeyPair()
const key = await wcb.derivePassword({ privateKey, publicKey, length: 16 })
// ArrayBuffer {
//   [Uint8Contents]: <ac 54 d5 01 74 ca d6 87 f5 65 18 d0 4f e4 0f 18 77 ... more bytes>,
//   byteLength: 16
// }
```


### Fingerprinting
Methods for calculating fingerprints of public keys. Note that the fingerprints differ from the fingerprints from `ssh-keygen`.

#### `sha256Fingerprint`
Calculate a SHA-256 fingerprint of a key. It has a length of 64 hex chars.

```js
const { publicKey } = await wcb.generateKeyPair()
const fingerprintBits = await wcb.sha256Fingerprint(publicKey)
const fingerprint = wcb.encodeHex(fingerprintBits)
// aca8f766cdef8346177987a86b0f04b14fd4060b0e2478f941adc91982d6668c
```

#### `sha1Fingerprint`
Calculate a SHA-1 fingerprint of a key. It has a length of 40 hex chars.

```js
const { publicKey } = await wcb.generateKeyPair()
const fingerprintBits = await wcb.sha1Fingerprint(publicKey)
const fingerprint = wcb.encodeHex(fingerprintBits)
// d04f73b7eb0b865a8d4711b5a379273a27c65581
```


### Key Import and Export
Tools for exchanging keys. Also comes with convenient helpers to deal with PEM formatted keys.

#### `exportKey`
Exports aes key data as ArrayBuffer:

```js
const key = await wcb.generateKey()
const data = await wcb.exportKey(key)
// ArrayBuffer {
//   [Uint8Contents]: <ac 54 d5 01 74 ca d6 87 f5 65 18 d0 4f e4 0f 18 77 ... more bytes>,
//   byteLength: 32
// }
```

#### `importKey`
Import aes key data, returns CryptoKey:

```js
const data = new Uint8Array([
  210, 29, 179, 47, 204, 90, 109, 111, 95, 64, 50, 48, 192, 105, 44, 236,
  74, 120, 2, 193, 83, 122, 22, 99, 202, 73, 20, 23, 187, 160, 140, 112
])
const key = await wcb.importKey(data)
// CryptoKey {
//   type: 'secret',
//   extractable: true,
//   algorithm: { name: 'AES-GCM', length: 256 },
//   usages: [ 'encrypt', 'decrypt' ]
// }
```

#### `exportPublicKeyPem`
Utility function to export a public key as pem:

```js
const { publicKey } = await wcb.generateKeyPair()
const pem = await wcb.exportPublicKeyPem(publicKey)
// -----BEGIN PUBLIC KEY-----
// MIGbMBAGByqGSM49AgEGBSuBBAAjA4GGAAQBdjbvVyj7NglRaHLqgn2l+Rcw1Lev
// 50/9xL6qvIwYSv84jU6xLOMJGY7nouU0tuWmm1+ojHsd3raDfxjsNSmKSOEAbSFJ
// 1rnRvBU/DflEh0i/RofX4vmKH3quCKPQ8T1NQoQijyKEOjkQDFqDgpPW03SMusqs
// d9/kSDNOMLm+EAIA6C0=
// -----END PUBLIC KEY-----
```

#### `importPublicKeyPem`
Given a pem of a public key, returns the CryptoKey:

```js
const pem = `-----BEGIN PUBLIC KEY-----
MIGbMBAGByqGSM49AgEGBSuBBAAjA4GGAAQBdjbvVyj7NglRaHLqgn2l+Rcw1Lev
50/9xL6qvIwYSv84jU6xLOMJGY7nouU0tuWmm1+ojHsd3raDfxjsNSmKSOEAbSFJ
1rnRvBU/DflEh0i/RofX4vmKH3quCKPQ8T1NQoQijyKEOjkQDFqDgpPW03SMusqs
d9/kSDNOMLm+EAIA6C0=
-----END PUBLIC KEY-----`
const publicKey = await wcb.importPublicKeyPem(pem)
// CryptoKey {
//   type: 'public',
//   extractable: true,
//   algorithm: { name: 'ECDH', namedCurve: 'P-521' },
//   usages: []
// }
```

#### `exportPrivateKeyPem`
Utility function to export a private key as pem:

```js
const { privateKey } = await wcb.generateKeyPair()
const pem = await wcb.exportPrivateKeyPem(privateKey)
// -----BEGIN PRIVATE KEY-----
// MIHuAgEAMBAGByqGSM49AgEGBSuBBAAjBIHWMIHTAgEBBEIBcf8zEjlssqn4aTEB
// RR43ofwH/4BAXDAAd83Kz1Dyd+Ko0pit4ESgqSu/bJMdnDrpiGYuz0Klarwip8LD
// rYd9mEahgYkDgYYABAF2Nu9XKPs2CVFocuqCfaX5FzDUt6/nT/3Evqq8jBhK/ziN
// TrEs4wkZjuei5TS25aabX6iMex3etoN/GOw1KYpI4QBtIUnWudG8FT8N+USHSL9G
// h9fi+Yofeq4Io9DxPU1ChCKPIoQ6ORAMWoOCk9bTdIy6yqx33+RIM04wub4QAgDo
// LQ==
// -----END PRIVATE KEY-----
```

#### `importPrivateKeyPem`
Given a pem of a private key, returns the CryptoKey:

```js
const pem = `-----BEGIN PRIVATE KEY-----
MIHuAgEAMBAGByqGSM49AgEGBSuBBAAjBIHWMIHTAgEBBEIBcf8zEjlssqn4aTEB
RR43ofwH/4BAXDAAd83Kz1Dyd+Ko0pit4ESgqSu/bJMdnDrpiGYuz0Klarwip8LD
rYd9mEahgYkDgYYABAF2Nu9XKPs2CVFocuqCfaX5FzDUt6/nT/3Evqq8jBhK/ziN
TrEs4wkZjuei5TS25aabX6iMex3etoN/GOw1KYpI4QBtIUnWudG8FT8N+USHSL9G
h9fi+Yofeq4Io9DxPU1ChCKPIoQ6ORAMWoOCk9bTdIy6yqx33+RIM04wub4QAgDo
LQ==
-----END PRIVATE KEY-----`
const privateKey = await wcb.importPrivateKeyPem(pem)
// CryptoKey {
//   type: 'private',
//   extractable: true,
//   algorithm: { name: 'ECDH', namedCurve: 'P-521' },
//   usages: [ 'deriveKey' ]
// }
```

#### `exportEncryptedPrivateKeyPem`
Encrypt a private key with passphrase and export as PEM:

```js
const { privateKey } = await wcb.generateKeyPair()
const passphrase = 'secure'
const pem = await wcb.exportEncryptedPrivateKeyPem({ key: privateKey, passphrase })
// -----BEGIN ENCRYPTED PRIVATE KEY-----
// MIIBZjBgBgkqhkiG9w0BBQ0wUzAyBgkqhkiG9w0BBQwwJQQQi9FqU3dish14EV99
// Bz3tugIDAPoAMAwGCCqGSIb3DQIJBQAwHQYJYIZIAWUDBAEqBBBK8mQ193fArwsz
// PKt2SvCkBIIBAOKBG4NQBWNDUUxSTJAMy5XnOU6nnX+Sisb9uu8/bAhxRtn3ItTo
// vGCs2MxtTKQhBRC5WdjU7oEe5rZAsWfoYdb567hPnl19QRaf2cneTNHT5qDdzRF+
// PrLRyw2+XUDEeeU5vhC4E29LZgigeYdSd2r8fOOcJxrKmMzkyFJCrYFhoqpdw2IS
// 4FQgJ9axQ2AncSaTqbuhBFQcoIFMrJ21ncVeEtTHS3428RHQJF1czNb/qnj/uIg7
// ta5OqDeXseEgnF+StrDcnSjkSuqMqXVeKsduZd/5JZz25sbHgLBzRgiqf/1jyrrl
// j8CC5qXX2PQH6RKBTys/1fRY++y3OVALhb8=
// -----END ENCRYPTED PRIVATE KEY-----
```

#### `importEncryptedPrivateKeyPem`
Decrypt passphrase encrypted private key pem and import the key:

```js
const pem = `-----BEGIN ENCRYPTED PRIVATE KEY-----
MIIBZjBgBgkqhkiG9w0BBQ0wUzAyBgkqhkiG9w0BBQwwJQQQi9FqU3dish14EV99
Bz3tugIDAPoAMAwGCCqGSIb3DQIJBQAwHQYJYIZIAWUDBAEqBBBK8mQ193fArwsz
PKt2SvCkBIIBAOKBG4NQBWNDUUxSTJAMy5XnOU6nnX+Sisb9uu8/bAhxRtn3ItTo
vGCs2MxtTKQhBRC5WdjU7oEe5rZAsWfoYdb567hPnl19QRaf2cneTNHT5qDdzRF+
PrLRyw2+XUDEeeU5vhC4E29LZgigeYdSd2r8fOOcJxrKmMzkyFJCrYFhoqpdw2IS
4FQgJ9axQ2AncSaTqbuhBFQcoIFMrJ21ncVeEtTHS3428RHQJF1czNb/qnj/uIg7
ta5OqDeXseEgnF+StrDcnSjkSuqMqXVeKsduZd/5JZz25sbHgLBzRgiqf/1jyrrl
j8CC5qXX2PQH6RKBTys/1fRY++y3OVALhb8=
-----END ENCRYPTED PRIVATE KEY-----`
const passphrase = 'secure'
const privateKey = await wcb.importEncryptedPrivateKeyPem({ pem, passphrase })
// CryptoKey {
//   type: 'private',
//   extractable: true,
//   algorithm: { name: 'ECDH', namedCurve: 'P-521' },
//   usages: [ 'deriveKey' ]
// }
```

#### `exportEncryptedPrivateKeyPemTo`
Encrypt a private key with key pair and export as PEM:

```js
const share = await wcb.generateKeyPair()
const alice = await wcb.generateKeyPair()
const bob = await wcb.generateKeyPair()
const pem = await wcb.exportEncryptedPrivateKeyPem({
  key: share.privateKey,
  privateKey: alice.privateKey,
  publicKey: bob.publicKey
})
// -----BEGIN ENCRYPTED PRIVATE KEY-----
// MIIBZjBgBgkqhkiG9w0BBQ0wUzAyBgkqhkiG9w0BBQwwJQQQPjQ7/lIPfbHQQHGi
// QqXaJgIDAPoAMAwGCCqGSIb3DQIJBQAwHQYJYIZIAWUDBAEqBBCI371u15dp+o/N
// Iqq+O3DQBIIBAJ+hnDVdydYcKYTmmhhUmwybqNkWEWi9pG4un5Xf7bEtm2A2qzoi
// 73XnmHPfXW+435RgMbLRtJOxqDa519kRvedO1nNIw1Iycs9GynTar+D+fBE/tFmJ
// 66XwlhKcKe0zMtSoi4FnkeyueEMYpJ7UDx+zVABqwwZdSFUrraeg6g/ljL1SGslg
// xDhTEyULoLyYV4G1+2t+rRXdzr408v6AAi+fJh/iBiwqd6clc+oNW0iXxHsi5/nH
// kETXf3RmXRonS7Ema+zhe3hMGlGGV1OMixaUZHGiIB6zc4fHHyzb5ippXEDkZ6e3
// U+l5Po65rsFkaAcDupfN18Ez9FRAyYbNkz8=
// -----END ENCRYPTED PRIVATE KEY-----
```

#### `importEncryptedPrivateKeyPemFrom`
Decrypts encrypted private key PEM with key pair:

```js
// alice and bob from above
const privateKey = await wcb.importEncryptedPrivateKeyPemFrom({
  pem,
  privateKey: bob.privateKey,
  publicKey: alice.publicKey
})
// CryptoKey {
//   type: 'private',
//   extractable: true,

CHANGELOG.md
# Changelog

## v4.0.0 - Preselected Cipher
Went full circle. Back to a pure collection of functions. Cipher is set to ECDH P-521 AES-256-CBC.

**Features:**
* initialization vector included in cipher, no manual iv generation needed anymore
* import and export encrypted private key pems

**Breaking changes:**
* all functions previously under `utils` are now toplevel exports
* no more `Webcryptobox` class, class methods moved to toplevel exports using preselected cipher
* `derivePrivateKey` has been renamed to `getPrivateKey`


## v3.0.0 - Configurable Cipher
A restructured Webcryptobox, with configurable cipher. The export is now a class, which has to be initialized.

**Features:**
* configurable ciphers:
  - choose between `P-256`, `P-384` and `P-521` ec curves
  - choose a aes cipher: `GCM` or `CBC`
  - choose aes key length: `128` or `256`
* new functions `importKey` and `exportKey` to exchange aes keys

**Breaking changes:**
* decode and encode functions have been moved to `utils` export:
  - `decodeText`
  - `encodeText`
  - `decodeHex`
  - `encodeHex`
  - `decodeBase64`
  - `encodeBase64`
* `Webcryptobox` export is now a class, which is instantiated with `curve`, `mode` and `length` parameters configuring the ciphers
* `generateSha1Fingerprint` has been renamed to `sha1Fingerprint`
* `generateSha256Fingerprint` has been renamed to `sha256Fingerprint`

**Fix:**
* PEM wrapping has been set to 64 chars according to the standard


## v2.0.0 - SHA-1 Fingerprints
Introduction of a SHA-1 fingerprint function: `generateSha1Fingerprint`.

**Breaking change:**
* rename `generateFingerprint` to `generateSha256Fingerprint`


## v1.1.0 - Wrap PEM Lines
Small improvement: break lines after 80 chars in PEMs.


## v1.0.2 - README Typos
Fixes typo in example on README.


## v1.0.1 - Tight Package
Tightened the npm package by removing non-essential files.


## v1.0.0 - Hello World!
Initial release of Webcryptobox


webcryptobox.js
// load the webcrypto api either from `window.crypto` when in browser
// or import from crypto standard lib when in node
const crypto = typeof window === 'undefined' ? (await import('crypto')).webcrypto : window.crypto


// crypto config
const EC_PARAMS = {
  name: 'ECDH',
  namedCurve: 'P-521'
}
const AES_PARAMS = {
  name: `AES-CBC`,
  length: 256
}
const IV_LENGTH = 16
const PBKDF2_PARAMS = {
  name: 'PBKDF2',
  iterations: 64000,
  hash: 'SHA-256'
}
const OIDS = {
  pbkdf2: '06092a864886f70d01050c', // PBKDF2
  pbes2: '06092a864886f70d01050d',  // PBES2
  hash: '06082a864886f70d02090500', // SHA-256
  cipher: '060960864801650304012a'  // AES-256-CBC
}

export const cipher = `${EC_PARAMS.name}-${EC_PARAMS.namedCurve}-${AES_PARAMS.name.split('-')[0]}-${AES_PARAMS.length}-${AES_PARAMS.name.split('-')[1]}`


// utils

// decode text message
export const decodeText = text => {
  const enc = new TextEncoder()
  return enc.encode(text)
}

// encode data as text
export const encodeText = data => {
  const dec = new TextDecoder()
  return dec.decode(data)
}


// decode hex string
export const decodeHex = hex => new Uint8Array(
  hex.match(/[\da-f]{2}/gi).map(h => parseInt(h, 16))
)

// encode data as hex string
export const encodeHex = data => Array.from(new Uint8Array(data))
  .map(x => ('00' + x.toString(16)).slice(-2))
  .join('')


// decode base64 string
export const decodeBase64 = base64 => {
  var i, d = atob(base64.trim()), b = new Uint8Array(d.length)
  for (i = 0; i < d.length; i++) b[i] = d.charCodeAt(i)
  return b
}

// encode data as base64 string
export const encodeBase64 = data => btoa(String.fromCharCode(...new Uint8Array(data)))


// pem

// decode public ecdh key pem
const decodePublicKeyPem = pem => decodeBase64(pem.trim().slice(27, -25).replace(/\n/g, ''))

// encode public ecdh key as pem
const encodePublicKeyPem = data => `-----BEGIN PUBLIC KEY-----
${encodeBase64(data).match(/.{1,64}/g).join('\n')}
-----END PUBLIC KEY-----`

// decode private key pem
const decodePrivateKeyPem = pem => decodeBase64(pem.trim().slice(28, -26).replace(/\n/g, ''))

// encode private key as pem
const encodePrivateKeyPem = data => `-----BEGIN PRIVATE KEY-----
${encodeBase64(data).match(/.{1,64}/g).join('\n')}
-----END PRIVATE KEY-----`


// encrypted pem encoding

// utilities
const decimalToHex = d => {
   const h = (d).toString(16)
   return h.length % 2 ? '0' + h : h
}
const hexLength = d => {
   const h = (d.length / 2).toString(16)
   return h.length % 2 ? '0' + h : h
}

// encode encrypted private key as pem
const encodeEncryptedPrivateKeyPem = ({ wrappedKey, iv, salt }) => {
  const wrappedKeyHex = encodeHex(wrappedKey)
  const saltHex = encodeHex(salt)
  const ivHex = encodeHex(iv)
  const iter = '00' + decimalToHex(PBKDF2_PARAMS.iterations)
  const iterInteger = '02' + decimalToHex(iter.length / 2) + iter
  const saltOctet = '04' + hexLength(saltHex) + saltHex
  const ivOctet = '04' + hexLength(ivHex) + ivHex
  const keyOctetPadding = hexLength(wrappedKeyHex).length / 2 === 2 ? '82' : '81'
  const keyOctet = '04' + keyOctetPadding + hexLength(wrappedKeyHex) + wrappedKeyHex
  const aesContainer = '30' + hexLength(OIDS.cipher + ivOctet)
  const hashContainer = '30' + hexLength(OIDS.hash)
  const pbkdf2InnerParameters = saltOctet + iterInteger + hashContainer + OIDS.hash
  const pbkdf2InnerContainer = '30' + hexLength(pbkdf2InnerParameters)
  const pbkdf2Parameters = OIDS.pbkdf2 + pbkdf2InnerContainer + pbkdf2InnerParameters
  const pbkdf2Container = '30' + hexLength(pbkdf2Parameters)
  const pbes2InnerParameters = pbkdf2Container + pbkdf2Parameters + aesContainer + OIDS.cipher + ivOctet
  const pbes2InnerContainer = '30' + hexLength(pbes2InnerParameters)
  const sequenceParameters = OIDS.pbes2 + pbes2InnerContainer + pbes2InnerParameters
  const sequenceContainer = '30' + hexLength(sequenceParameters) + sequenceParameters
  const sequenceLength = hexLength(sequenceContainer + keyOctet)
  const sequencePadding = sequenceLength.length / 2 === 2 ? '82' : '81'
  const sequence = '30' + sequencePadding + sequenceLength + sequenceContainer + keyOctet

  const asnKey = decodeHex(sequence)

  return `-----BEGIN ENCRYPTED PRIVATE KEY-----
${encodeBase64(asnKey).match(/.{1,64}/g).join('\n')}
-----END ENCRYPTED PRIVATE KEY-----`
}

const decodeEncryptedPrivateKeyPem = pem => {
  const pemData = decodeBase64(pem.trim().slice(38, -36).replace(/\n/g, ''))
  const hex = encodeHex(pemData)

  if (!hex.includes(OIDS.pbkdf2) || !hex.includes(OIDS.pbes2)) {
    throw(new Error('Invalid pem: no PBKDF2 or PBES2 header'))
  }
  if (!hex.includes(OIDS.cipher)) {
    throw(new Error('Invalid pem: unsupported cipher'))
  }
  if (!hex.includes(OIDS.hash)) {
    throw(new Error('Invalid pem: unsupported hash'))
  }

  const saltBegin = hex.indexOf(OIDS.pbkdf2) + 28
  const ivBegin = hex.indexOf(OIDS.cipher) + 24
  const saltLength = parseInt(hex.substr(saltBegin, 2), 16)
  const ivLength = parseInt(hex.substr(ivBegin, 2), 16)
  const saltHex = hex.substr(saltBegin + 2, saltLength * 2)
  const ivHex = hex.substr(ivBegin + 2, ivLength * 2)
  const iterBegin = saltBegin + 4 + (saltLength * 2)
  const iterLength = parseInt(hex.substr(iterBegin, 2), 16)
  const iter = parseInt(hex.substr(iterBegin + 2, iterLength * 2), 16)
  const sequencePadding = hex.substr(2, 2) === '81' ? 8 : 10
  const parametersPadding = hex.substr(2, 2) === '81' ? 12 : 16
  const sequenceLength = parseInt(hex.substr(sequencePadding, 2), 16)
  const encryptedDataBegin = parametersPadding + (sequenceLength * 2)
  const encryptedDataPadding = hex.substr(encryptedDataBegin - 2, 2) === '81' ? 2 : 4
  const encryptedDataLength = parseInt(hex.substr(encryptedDataBegin, 6), 16)
  const encryptedData = hex.substr(encryptedDataBegin + encryptedDataPadding, (encryptedDataLength * 2))

  const salt = decodeHex(saltHex)
  const iv = decodeHex(ivHex)
  const wrappedKey = decodeHex(encryptedData)

  return {
    wrappedKey,
    iv,
    salt
  }
}


// key generation

// generate ecdh key pair pair
export const generateKeyPair = () => crypto.subtle.generateKey(
  EC_PARAMS,
  true,
  ['deriveKey', 'deriveBits']
)

// generate aes key
export const generateKey = () => crypto.subtle.generateKey(
  AES_PARAMS,
  true,
  ['encrypt', 'decrypt']
)

// derive aes encryption key from ecdh public and private key
export const deriveKey = ({ publicKey, privateKey }) => crypto.subtle.deriveKey(
  {
    ...EC_PARAMS,
    public: publicKey
  },
  privateKey,
  AES_PARAMS,
  true,
  ['encrypt', 'decrypt']
)

// derive bits from ecdh key pair
export const deriveBits = ({ length, publicKey, privateKey }) => crypto.subtle.deriveBits(
  {
    ...EC_PARAMS,
    public: publicKey
  },
  privateKey,
  length * 8
)

// derive wrapping key from passphrase with pkdf2
const deriveWrappingKey = async ({ salt, passphrase }) => {
  const baseKey = await crypto.subtle.importKey(
    'raw',
    passphrase,
    {
      name: 'PBKDF2'
    },
    false,
    ['deriveKey']
  )
  
  return crypto.subtle.deriveKey(
    {
      ...PBKDF2_PARAMS,
      salt
    },
    baseKey,
    AES_PARAMS,
    false,
    ['wrapKey', 'unwrapKey']
  )
}

// wrap a key in pkcs8 encrypted with wrappingKey
const wrapKey = async ({ key, wrappingKey }) => {
  const iv = await crypto.getRandomValues(new Uint8Array(16))
  const wrappedKey = await crypto.subtle.wrapKey(
    'pkcs8',
    key,
    wrappingKey,
    {
      ...AES_PARAMS,
      iv: iv
    }
  )

  return {
    wrappedKey,
    iv
  }
}

// unwrap a key from pkcs8 encrypted with wrappingKey
const unwrapKey = async ({ wrappedKey, wrappingKey, iv }) => {
  return crypto.subtle.unwrapKey(
    'pkcs8',
    wrappedKey,
    wrappingKey,
    {
      ...AES_PARAMS,
      iv: iv
    },
    EC_PARAMS,
    true,
    ['deriveKey', 'deriveBits']
  )
}


// derive password from private and public key
export const derivePassword = async ({ privateKey, publicKey, length }) => {
  const aes_length = AES_PARAMS.length / 8
  const bits = await deriveBits({ privateKey, publicKey, length: length + aes_length })
  return bits.slice(aes_length)
}

// derive public key from private key
export const getPublicKey = async privateKey => {
  const jwk = await crypto.subtle.exportKey(
    'jwk',
    privateKey
  )
  delete jwk.d
  return crypto.subtle.importKey(
    'jwk',
    jwk,
    EC_PARAMS,
    true,
    []
  )
}


// fingerprinting

// calculate a sha256 fingerprint of a key
export const sha256Fingerprint = async key => {
  const keyBits = await crypto.subtle.exportKey('spki', key)
  return crypto.subtle.digest('SHA-256', keyBits)
}

// calculate a sha1 fingerprint of a key
export const sha1Fingerprint = async key => {
  const keyBits = await crypto.subtle.exportKey('spki', key)
  return crypto.subtle.digest('SHA-1', keyBits)
}


// import & export keys

// import a raw aes key
export const importKey = data => crypto.subtle.importKey(
  'raw',
  data,
  AES_PARAMS,
  true,
  ['encrypt', 'decrypt']
)

// export an aes key as raw bytes
export const exportKey = key => crypto.subtle.exportKey(
  'raw',
  key
)

// internal functions

// export a public ecdh key as spki
const exportPublicKey = publicKey => crypto.subtle.exportKey(
  'spki',
  publicKey
)

// export a private ecdh key as pkcs8
const exportPrivateKey = privateKey => crypto.subtle.exportKey(
  'pkcs8',
  privateKey
)

// import a public ecdh key spki
const importPublicKey = data => crypto.subtle.importKey(
  'spki',
  data,
  EC_PARAMS,
  true,
  []
)

// import a private ecdh key pkcs8
const importPrivateKey = data => crypto.subtle.importKey(
  'pkcs8',
  data,
  EC_PARAMS,
  true,
  ['deriveKey', 'deriveBits']
)


// export public key as pem
export const exportPublicKeyPem = async key => {
  const data = await exportPublicKey(key)
  return encodePublicKeyPem(data)
}

// import public key pem
export const importPublicKeyPem = pem => {
  const data = decodePublicKeyPem(pem)
  return importPublicKey(data)
}

// export private key as pem
export const exportPrivateKeyPem = async key => {
  const data = await exportPrivateKey(key)
  return encodePrivateKeyPem(data)
}

// import private key pem
export const importPrivateKeyPem = pem => {
  const data = decodePrivateKeyPem(pem)
  return importPrivateKey(data)
}

// export encrypted private key as pem
export const exportEncryptedPrivateKeyPem = async ({ key, passphrase }) => {
  const salt = await crypto.getRandomValues(new Uint8Array(16))
  const wrappingKey = await deriveWrappingKey({ salt, passphrase })
  const { wrappedKey, iv } = await wrapKey({ key, wrappingKey })
  return encodeEncryptedPrivateKeyPem({ wrappedKey, iv, salt })
}

// import encrypted private key pem
export const importEncryptedPrivateKeyPem = async ({ pem, passphrase }) => {
  const { wrappedKey, iv, salt } = decodeEncryptedPrivateKeyPem(pem)
  const wrappingKey = await deriveWrappingKey({ salt, passphrase })
  return unwrapKey({ wrappedKey, wrappingKey, iv })
}


// derive wrapping key from peer
const deriveWrappingKeyFrom = async ({ salt, privateKey, publicKey }) => {
  const passphraseBits = await derivePassword({ privateKey, publicKey, length: 32 })
  const passphraseHex = encodeHex(passphraseBits)
  const passphrase = decodeText(passphraseHex)
  return deriveWrappingKey({ salt, passphrase })
}

// export encrypted private key as pem to peer
export const exportEncryptedPrivateKeyPemTo = async ({ key, privateKey, publicKey }) => {
  const salt = await crypto.getRandomValues(new Uint8Array(16))
  const wrappingKey = await deriveWrappingKeyFrom({ salt, privateKey, publicKey })
  const { wrappedKey, iv } = await wrapKey({ key, wrappingKey })
  return encodeEncryptedPrivateKeyPem({ wrappedKey, iv, salt })
}

// import encrypted private key pem from peer
export const importEncryptedPrivateKeyPemFrom = async ({ pem, privateKey, publicKey }) => {
  const { wrappedKey, iv, salt } = decodeEncryptedPrivateKeyPem(pem)
  const wrappingKey = await deriveWrappingKeyFrom({ salt, privateKey, publicKey })
  return unwrapKey({ wrappedKey, wrappingKey, iv })
}


// encryption & decryption

// internal: generate a initialization vector (aka nonce)
const generateIv = () => crypto.getRandomValues(new Uint8Array(IV_LENGTH))


// symmetric encryption

// encrypt a message with aes key
export const encrypt = async ({ message, key }) => {
  const iv = generateIv()
  const encrypted = await crypto.subtle.encrypt(
    {
      ...AES_PARAMS,
      iv
    },
    key,
    message
  )
  const box = new Uint8Array(IV_LENGTH + encrypted.byteLength)
  box.set(iv, 0)
  box.set(new Uint8Array(encrypted), IV_LENGTH)
  return box
}

// decrypt a message with aes key
export const decrypt = async ({ box, key }) => {
  const iv = box.slice(0, IV_LENGTH)
  const encrypted = box.slice(IV_LENGTH)
  const message = await crypto.subtle.decrypt(
    {
      ...AES_PARAMS,
      iv
    },
    key,
    encrypted
  )
  return new Uint8Array(message)
}


// asymmetric encryption

// encrypt with private key and public public peer key
export const encryptTo = async ({ message, privateKey, publicKey }) => {
  const key = await deriveKey({ privateKey, publicKey })
  return encrypt({ message, key })
}

// decrypt with private key and public peer key
export const decryptFrom = async ({ box, privateKey, publicKey }) => {
  const key = await deriveKey({ privateKey, publicKey })
  return decrypt({ box, key })
}
